[iOS] Add iOS test app skeleton

xta0 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26261 [iOS] Add iOS test app skeleton**

### Summary

Previously we have enabled the CI jobs for Pull Requests and nightly build

-  **#25840 [iOS][Circle CI] Add PR jobs for iOS builds**
-  **#26074 [IOS][CIRCLE CI] Nightly jobs for iOS builds**

The testing phase is missing in the nightly build process.  Although we are able to generate the build and upload it to the AWS,  there is no way to know whether the binary is valid or not (there could be a linking error). To add the test phase to the process, we need 

1. Put a dummy test App in the repo.
2. After the build jobs finishes, manually links the static libs to the dummy app to produce an executable using the XCode tool chain.
3. If there is no linking error, then upload the binaries to AWS. If there is an error, then stops the following process and reports an error in CI.

The second and third steps depends on the first step which needs to be landed first. 

### Test Plan
- Don't break any existing CI jobs

Differential Revision: [D17408929](https://our.internmc.facebook.com/intern/diff/D17408929)

xta0 committed:
[iOS] Add iOS tset app skeleton

xta0 committed:
Update on "[iOS] Add iOS tset app skeleton"

shoumikhin commented on ios/TestApp/TestApp/AppDelegate.h:
```diff
@@ -0,0 +1,8 @@
+#import <UIKit/UIKit.h>
+
```
nit:
Please add NS_ASSUME_NONNULL macros in all headers

shoumikhin commented on ios/TestApp/TestApp/AppDelegate.h:
```diff
@@ -0,0 +1,8 @@
+
+@interface AppDelegate : UIResponder <UIApplicationDelegate>
+
+@property (strong, nonatomic) UIWindow *window;
```
nit: `strong` is default

shoumikhin commented on ios/TestApp/TestApp/AppDelegate.m:
```diff
@@ -0,0 +1,43 @@
+#import "AppDelegate.h"
+
+@interface AppDelegate ()
```
nit: not needed?

shoumikhin commented on ios/TestApp/TestApp/AppDelegate.m:
```diff
@@ -0,0 +1,43 @@
+@implementation AppDelegate
+
+
+- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
```
nit: may like to remove everything in this file except of this method

shoumikhin commented on ios/TestApp/TestApp/Base.lproj/Main.storyboard:
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
nit: ideally we want to build UI in code to be able to review it, so any storyboards other than LaunchScreen should be deprecated

shoumikhin commented on ios/TestApp/TestApp/ViewController.mm:
```diff
@@ -0,0 +1,16 @@
+#import "ViewController.h"
+#import <LibTorch/LibTorch.h>
+
+@interface ViewController ()
```
nit: not needed?

shoumikhin commented on ios/TestApp/TestApp/main.m:
```diff
@@ -0,0 +1,8 @@
+#import <UIKit/UIKit.h>
```
nit: UIKit import is redundant, since it is already imported by AppDelegate anyway

ezyang commented:
Hi @xta0, you assigned me as a reviewer on this PR. Is there anything in particular you would like me to review? I am not familiar with iOS so cannot help with those parts ;)

xta0 commented:
> Hi @xta0, you assigned me as a reviewer on this PR. Is there anything in particular you would like me to review? I am not familiar with iOS so cannot help with those parts ;)

Hey @ezyang, please feel free to ignore the iOS code.  Just want to sync with you that this one is part of the iOS CI process, and the testing approach is different from the existing nightly jobs.

xta0 committed:
Update on "[iOS] Add iOS test app skeleton"


### Summary

Previously we have enabled the CI jobs for Pull Requests and nightly build

-  **#25840 [iOS][Circle CI] Add PR jobs for iOS builds**
-  **#26074 [IOS][CIRCLE CI] Nightly jobs for iOS builds**

The testing phase is missing in the nightly build process.  Although we are able to generate the build and upload it to the AWS,  there is no way to know whether the binary is valid or not (there could be a linking error). To add the test phase to the process, we need 

1. Put a dummy test App in the repo.
2. After the build jobs finishes, manually link the static libs to the dummy app to produce an executable using the xcode tool chain.
3. If there is no linking error, then upload the binaris to AWS. If there is an error, then stops the following process and reports an error in CI.

The second and third steps depends on the first step which needs to be landed first. 

### Test Plan
- Don't break any existing CI jobs

ezyang commented:
FTR I reviewed the other diff.

rohan-varma commented:
I think this is breaking a CI job that checks that no tabs are used: https://dev.azure.com/pytorch/PyTorch/_build/results?buildId=11551&view=logs
